File: leetcode/0015-3Sum.cc
// https://leetcode.com/problems/3sum/
// Medium

#include<vector>
#include<utility>
#include<unordered_map>
#include<unordered_set>
#include<algorithm>
using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> ret;
        sort(nums.begin(), nums.end());
        
        int size = nums.size();

        for (int i = 0; i < size;) {
            int lo = i+1;
            int hi = size-1;

            while (lo < hi) {
                int sum = nums[lo] + nums[hi] + nums[i];
                if (sum < 0) {
                    ++lo;
                } else if (sum > 0) {
                    --hi;
                } else {
                    ret.push_back({nums[i], nums[lo], nums[hi]});
                    ++lo;
                    while (lo < hi && nums[lo] == nums[lo-1]) ++lo;
                }
            }
            ++i;
            while (i > 0 && i < size && nums[i] == nums[i-1]) ++i;
        }
        return ret;
    }
};
class Solution1 {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> v;
        sort(nums.begin(), nums.end());
        int size = nums.size();
        for (int i = 0; i < size - 1; ++i) {
            int left = i+1, right = size - 1, sum = -nums[i];
            while (left < right) {
                // cout << left << ' ' << right << endl;
                if (nums[left] + nums[right] > sum) {
                    --right;
                } else if (nums[left] + nums[right] < sum) {
                    ++left;
                } else {
                    v.push_back(vector<int> {nums[i], nums[left], nums[right]});
                    int temp = left;
                    // cout << "I: " << i << endl;
                    // cout << left << ' ' << right << endl;
                    ++temp;
                    if (temp+1 < size && nums[temp] == nums[left]) {
                        while (1) {
                            if (temp+1 >= size || nums[temp] != nums[left]) {
                                // i = temp - 1;
                                // sum = -nums[i];
                                break;
                            }
                            ++temp;
                        }
                    }
                    
                    left = temp;
                    // cout << "I: " << i << endl;
                    // cout << left << ' ' << right << endl;
                }
            }
            if (i+1 < size && nums[i+1] == nums[i]) {
                while(1) {
                    ++i;
                    if (i+1 >= size || nums[i+1] != nums[i]) break;
                }
            }
        }
        return v;
    }
};

File: leetcode/python/0015-3sum.py
# https://leetcode.com/problems/3sum/
# Medium

# Two Sum Solution
# Quicker and more memory efficient
class SolutionTwoSum:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        def twoSum(nums: List[int], target: int, start: int) -> List[List[int]]:
            lo, hi = start, len(nums)-1
            ret = []
            while lo < hi:
                if nums[lo]+nums[hi]+target > 0:
                    hi -= 1
                elif nums[lo]+nums[hi]+target < 0:
                    lo += 1
                else:
                    ret.append([nums[lo], nums[hi], target])
                    lo += 1
                    while lo < hi and nums[lo] == nums[lo-1]:
                        lo += 1
            return ret
        
        nums.sort()
        ret = []
        for i, n in enumerate(nums):
            if n > 0:
                break
            if i == 0 or nums[i] > nums[i-1]:
                ret += twoSum(nums, n, i+1) 

        return ret

# Set solution
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        d = {}
        ret = []
        for i, n in enumerate(nums):
            d[n] = i
        
        for i in range(len(nums)):
            if i > 0 and nums[i] == nums[i-1]:
                continue
            for j in range(i+1, len(nums)):
                if j > i+1 and nums[j] == nums[j-1]:
                    continue
                find = -(nums[i] + nums[j])
                if find in d and d[find] > j:
                    ret.append([nums[i], nums[j], find])

        return ret